File: 99. Recover Binary Search Tree/recoverTree.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
const recoverTree = (root) => {
  let prev = null;
  let first = null;
  let second = null;

  const inorder = (node) => {
    if (!node) {
      return;
    }

    inorder(node.left);

    if (prev && prev.val > node.val) {
      if (!first) {
        first = prev;
      }
      second = node;
    }
    prev = node;

    inorder(node.right);
  }

  inorder(root);

  [first.val, second.val] = [second.val, first.val]

  return root;
};

const root = new TreeNode(1)
root.left = new TreeNode(3)
root.left.right = new TreeNode(2)

console.log('recoverTree: ', recoverTree(root))

File: 99. Recover Binary Search Tree/recoverTree.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val = 0, left = None, right = None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def recoverTree(self, root: Optional[TreeNode]) -> None:
    """
    Do not return anything, modify root in-place instead.
    """
    def inorder(node):
      if not node:
        return

      inorder(node.left)

      if self.prev and self.prev.val > node.val:
        if not self.first:
          self.first = self.prev
        self.second = node
      self.prev = node

      inorder(node.right)
    
    self.prev = None
    self.first = None
    self.second = None

    inorder(root)

    self.first.val, self.second.val = self.second.val, self.first.val
  
    return root


root = TreeNode(1)
root.left = TreeNode(3)
root.left.right = TreeNode(2)

solution = Solution()
print('recoverTree: ', solution.recoverTree(root).val)
